libnavigation-base/src/main/java/com/mapbox/navigation/base/geodeeplink/GeoDeeplink.kt

```diff
@@ -0,0 +1,48 @@
+package com.mapbox.navigation.base.geodeeplink
+
+import com.mapbox.geojson.Point
+
+/**
+ * @param deeplink string deeplink value used to create this object
+ * @param point location coordinates specified by the deeplink, null when unknown
+ * @param placeQuery human readable description of the place
+ */
+class GeoDeeplink internal constructor(
+    val deeplink: String,
+    val point: Point?,
+    val placeQuery: String?
+) {
+
+    /**
+     * Indicates whether some other object is "equal to" this one.
+     */
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as GeoDeeplink
+
+        if (deeplink != other.deeplink) return false
+        if (point != other.point) return false
+        if (placeQuery != other.placeQuery) return false
+
+        return true
+    }
+
+    /**
+     * Returns a hash code value for the object.
+     */
+    override fun hashCode(): Int {
+        var result = deeplink.hashCode()
+        result = 31 * result + (point?.hashCode() ?: 0)
+        result = 31 * result + (placeQuery?.hashCode() ?: 0)
+        return result
+    }
+
+    /**
+     * Returns a string representation of the object.
+     */
+    override fun toString(): String {
+        return "GeoDeeplink(deeplink='$deeplink', point=$point, placeQuery=$placeQuery)"
+    }
```

Comment:
Will kotlin generate the same implementation if you make `GeoDeeplink` a `data class` ?

Reply:
`data class` has a [copy](https://kotlinlang.org/docs/data-classes.html#copying) function. Which allows you to corrupt the object elsewhere.

we have chosen to make `class Foo internal constructor` with overridden equals, hashCode and toString; for this reason alone. So I'm just being consistent with the sdk ([RouteProgress](https://github.com/mapbox/mapbox-navigation-android/blob/84a54835dfd606bf6ff8988b46db0b9b0e203c4e/libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteProgress.kt#L40) for example)

Jake says we should have a [`value class`](https://jakewharton.com/public-api-challenges-in-kotlin).

Reply:
you can add ➕ in the issue https://youtrack.jetbrains.com/issue/KT-11914